js/init_and_modules_test.go

```diff
@@ -0,0 +1,110 @@
+package js_test
+
+import (
+	"context"
+	"net/url"
+	"testing"
+	"time"
+
+	"github.com/loadimpact/k6/js"
+	"github.com/loadimpact/k6/js/common"
+	"github.com/loadimpact/k6/js/internal/modules"
+	"github.com/loadimpact/k6/lib"
+	"github.com/loadimpact/k6/lib/testutils"
+	"github.com/loadimpact/k6/loader"
+	"github.com/loadimpact/k6/stats"
+	"github.com/spf13/afero"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"gopkg.in/guregu/null.v3"
+)
+
+type CheckModule struct {
+	t             testing.TB
+	initCtxCalled int
+	vuCtxCalled   int
+}
+
+func (cm *CheckModule) InitCtx(ctx context.Context) {
+	cm.initCtxCalled++
+	assert.NotNil(cm.t, common.GetRuntime(ctx))
+	assert.NotNil(cm.t, common.GetInitEnv(ctx))
+	assert.Nil(cm.t, lib.GetState(ctx))
+}
+
+func (cm *CheckModule) VuCtx(ctx context.Context) {
+	cm.vuCtxCalled++
+	assert.NotNil(cm.t, common.GetRuntime(ctx))
+	assert.Nil(cm.t, common.GetInitEnv(ctx))
+	assert.NotNil(cm.t, lib.GetState(ctx))
+}
+
+func TestNewJSRunnerWithCustomModule(t *testing.T) {
+	checkModule := &CheckModule{t: t}
+	modules.Register("k6/check", checkModule)
+
+	script := `
+		var check = require("k6/check");
+		check.initCtx();
+
+		module.exports.options = { vus: 1, iteartions: 1 };
+		module.exports.default = function() {
+			check.vuCtx();
+		};
+	`
+
+	logger := testutils.NewLogger(t)
+	rtOptions := lib.RuntimeOptions{CompatibilityMode: null.StringFrom("base")}
+	runner, err := js.New(
+		logger,
+		&loader.SourceData{
+			URL:  &url.URL{Path: "blah", Scheme: "file"},
+			Data: []byte(script),
+		},
+		map[string]afero.Fs{"file": afero.NewMemMapFs(), "https": afero.NewMemMapFs()},
+		rtOptions,
+	)
+	require.NoError(t, err)
+	assert.Equal(t, checkModule.initCtxCalled, 1)
+	assert.Equal(t, checkModule.vuCtxCalled, 0)
+
+	vu, err := runner.NewVU(1, make(chan stats.SampleContainer, 100))
+	require.NoError(t, err)
+	assert.Equal(t, checkModule.initCtxCalled, 2)
+	assert.Equal(t, checkModule.vuCtxCalled, 0)
+
+	vuCtx, vuCancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer vuCancel()
+	activationParams := &lib.VUActivationParams{
+		RunContext:         vuCtx,
+		DeactivateCallback: func(lib.InitializedVU) {},
+	}
```

Comment:
You need to use separate activationParams between the archive and not archive run to not have a race .. I think

Reply:
hmm :thinking: I'll investigate later or on monday

Reply:
should be fixed in the last force-push